### PR TITLE
feat: `expectSafeMemory` & `expectSafeMemoryCall` cheatcodes

### PR DESCRIPTION
**Pull request by clabby**
# Overview

Adds two new cheatcodes:
```solidity
// Only allows memory writes to offsets [0x00, 0x60) ∪ [_min, _max) in the current subcontext. If any other
// memory is written to, the test will fail.
function expectSafeMemory(uint64 _min, uint64 _max) external;

// Only allows memory writes to offsets [0x00, 0x60) ∪ [_min, _max) in the next created subcontext.
// If any other memory is written to, the test will fail.
function expectSafeMemoryCall(uint64 _min, uint64 _max) external;
```

## TODO
- [x] Account for opcodes that can also expand memory through reading: `MLOAD`, `SHA3`, `CREATE`, `CREATE2`, `LOG0-4`

## Usage

- `expectSafeMemory` should be used when the developer wants to assert the memory safety of code within the given context, and it can be called multiple times in order to register several ranges that are okay to be touched.
- `expectSafeMemoryCall` should be used when the developer wants to assert the memory safety of code within the next created subcontext, and it can also be called multiple times in order to register several ranges that are okay to be touched.

**Note**: Both cheatcodes *implicitly* consider the memory range `[0x00, 0x60)` okay to modify. This range includes:
| Offset Range |     Description     |
|--------------|---------------------|
| [0x00, 0x40) | Scratch space       |
| [0x40, 0x60) | Free Memory Pointer |

We may want to consider allowing the developer to also lock this memory, but I can't think of many use-cases for this.

## Motivation

At the moment, assembloors have a hard time testing the memory safety of their functions that manually allocate / write to memory. A common tactic is to just check that the free memory pointer was updated correctly, but this ignores the test vector of erroneous / unintended writes to memory outside of the desired bounds.

## Solution

These new cheatcodes allow for asserting that a piece of code within the current subcontext or an entire subcontext only touches a given range (or ranges) of memory.

## Open Questions
cc @mattsse @bluealloy
1. Should we reset the ranges at a given context depth after that context has completed? Could do it in `call_end`.
2. ~~Is there a way to revert with data within the inspector's `step` function? At the moment, we're just reverting with `Return::Revert`, which doesn't give any information as to what went wrong.~~ Thanks matt!

### CODE REVIEW

**Review by mattsse — COMMENTED** [review]
I guess that makes sense,

wdyt @mds1 ?

**Comment by mds1** [comment]
This seems reasonable to me! One thought is that hardcoding support for `[0x00, 0x60)` is a solidity specific convention and I don't know if other languages also reserve those memory locations. So devs using foundry with other languages may want a way to turn off that default range. However since forge is geared towards solidity dev anyway, I'm ok having this as a default, and someone can PR in an overload or new cheats that don't default to using that range as the need arises.

> `expectSafeMemory` should be used when the developer wants to assert the memory safety of code within the given context, and it can be called multiple times in order to register several ranges that are okay to be touched.

So this would be in the test contract itself, e.g. if I'm testing library methods, and once called it applies to the rest of my test? 

> `expectSafeMemoryCall` should be used when the developer wants to assert the memory safety of code within the next created subcontext, and it can also be called multiple times in order to register several ranges that are okay to be touched.

And this applies to the next call only, but also includes all subsequent subcontexts created by that call, or just the top level call?

**Comment by clabby** [comment]
> This seems reasonable to me! One thought is that hardcoding support for `[0x00, 0x60)` is a solidity specific convention and I don't know if other languages also reserve those memory locations. So devs using foundry with other languages may want a way to turn off that default range. However since forge is geared towards solidity dev anyway, I'm ok having this as a default, and someone can PR in an overload or new cheats that don't default to using that range as the need arises.
> 
> > `expectSafeMemory` should be used when the developer wants to assert the memory safety of code within the given context, and it can be called multiple times in order to register several ranges that are okay to be touched.
> 
> So this would be in the test contract itself, e.g. if I'm testing library methods, and once called it applies to the rest of my test?
> 
> > `expectSafeMemoryCall` should be used when the developer wants to assert the memory safety of code within the next created subcontext, and it can also be called multiple times in order to register several ranges that are okay to be touched.
> 
> And this applies to the next call only, but also includes all subsequent subcontexts created by that call, or just the top level call?

For the first point, agree- Maybe we can add a configuration value in the `foundry.toml` to disable this range? Down to do this in a separate PR.

1. Yes
2. At the moment, this applies to the next subcontext *and* any future subcontexts that exist at the same depth (See the first open question, leaning towards clearing the requirements for a context at the end of a context's execution). The memory safety requirements are on a per-context-depth basis, so any subcontexts created by that call will not be checked.

**Comment by 0xPhaze** [comment]
What if memory is written to, but then restored to the original again. Does that count as a violation? Is the check performed after exiting the current context or after any operation that touches memory?

**Comment by clabby** [comment]
> What if memory is written to, but then restored to the original again? Does that count as a violation? Is the check performed after exiting the current context or after any operation that touches memory?

This does count as a violation, and the check is performed after any operation that alters memory. I'm down to alter the way that it works, but imo, even if you're writing and then restoring memory, it's best to express that you're doing so. What do you think?

**Comment by 0xPhaze** [comment]
Agree, though, you might also want to specifically use this to test that you are able to properly restore memory to a certain state. But this might be a slightly different use case and I imagine would make the code a bit more complex.

On another thought, does this handle the case when `call`, `delegatecall`, etc. specify memory output offset and size, but the call returns less data and doesn't overwrite the memory?
I.e. `call(gas(), 4, 0, 0, 0, 32)`

**Comment by clabby** [comment]
> Agree, though, you might also want to specifically use this to test that you are able to properly restore memory to a certain state. But this might be a slightly different use case and I imagine would make the code a bit more complex.
> 
> On another thought, does this handle the case when `call`, `delegatecall`, etc. specify memory output offset and size, but the call returns less data and doesn't overwrite the memory? I.e. `call(gas(), 4, 0, 0, 0, 32)`

Sorry- missed this in my notifs!
1. Agree- this is useful, but this cheatcode explicitly checks for memory modifications outside of a given range. You can test that memory was properly restored fairly easily in a test if it's not too many offsets that were modified and then restored as well, but it might be worth adding another cheatcode to make for better devex :smile: 
2. It does not. In your example, it'd be fine because the `CALL` opcode's params say it'll touch `[0x00, 0x20)` (which is allowed by the cheatcode by default), but in others, the cheatcode may throw. I'm open to adding support for this, but my feeling is that this cheatcode's main responsibility is to let the developer know when memory outside of an expected range is being mutated. For example:
    * If my memory size is `0x120`, my allowed range is `[0x00, 0x120)`, and I invoke `pop(call(gas(), 0x04, 0, 0, 0, 0x100, 0x20))`, no memory is modified (and if it were to be modified, it is within my allowed range). The cheatcode will not throw here.
    * If my memory size is `0x120`, my allowed range is `[0x00, 0x100)`, and I invoke `pop(call(gas(), 0x04, 0, 0, 0, 0x100, 0x20))`, no memory is modified, but the opcode's params say it wanted to touch `[0x100, 0x120)`. The cheatcode will throw here.
    * If my memory size is `0x80`, my allowed range is `[0x00, 0x100)`, and I invoke `pop(call(gas(), 0x04, 0, 0, 0, 0x100, 0x20))`, the cheatcode will throw because memory was expanded (even though no memory was overwritten).
  
For situations where you need to ensure that a call returns a certain data size (and thus actually writes the proper amount of data to the returndata offset supplied to the opcode), it may be easier to do so with a `returndatasize()` check rather than with a cheatcode.

**Comment by gakonst** [comment]
Supportive of the cheat code and excited for assembloors to get better testing...defer to you guys on the actual implementation although it looks good directionally on cursory glance.

**Comment by gakonst** [comment]
Btw the hooking on `step` can make things a bunch more expensive because it runs on every interpreter step, so want us to be sensitive or figure out ways on supporting these more expensive cheatcodes w/o impacting the whole test suite.

File: evm/src/executor/inspector/cheatcodes/mod.rs
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/mod.rs:389]: I'd like to extract this to a separate module or end of file, to keep the step fn lean.

File: evm/src/executor/inspector/cheatcodes/mod.rs
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/mod.rs:411]: I'd appreciate some docs here.

**Review by mattsse — CHANGES_REQUESTED** [review]
test look good, well documented cheat code impls.

only have style nits, otherwise lgtm.

pending @mds1

**Comment by mds1** [comment]
LGTM, only comment would be to track this in an issue:

> expectSafeMemory should be used when the developer wants to assert the memory safety of code within the given context, and it can be called multiple times in order to register several ranges that are okay to be touched.
>>  So this would be in the test contract itself, e.g. if I'm testing library methods, and once called it applies to the rest of my test?
>> > agree- Maybe we can add a configuration value in the foundry.toml to disable this range? Down to do this in a separate PR.

So just pulled it out into https://github.com/foundry-rs/foundry/issues/4578

**Comment by 0xPhaze** [comment]
Small UX nit: can the parameters be changed to `(uint256 _min, uint256 _max)` and be validated inside of foundry?

**Comment by mattsse** [comment]
haven't used this yet but seems reasonable ux request since more common?

I think we can do this @clabby ?

@0xPhaze please open a separate issue for this so we don't lose track

**Pull request merged**
